File: plugin/src/main/java/ru/privatenull/pnauth/transport/packetevents/PacketEventsDialogMapper.java
package ru.privatenull.pnauth.transport.packetevents;

import com.github.retrooper.packetevents.protocol.chat.clickevent.*;
import com.github.retrooper.packetevents.protocol.dialog.CommonDialogData;
import com.github.retrooper.packetevents.protocol.dialog.ConfirmationDialog;
import com.github.retrooper.packetevents.protocol.dialog.Dialog;
import com.github.retrooper.packetevents.protocol.dialog.MultiActionDialog;
import com.github.retrooper.packetevents.protocol.dialog.NoticeDialog;
import com.github.retrooper.packetevents.protocol.dialog.ServerLinksDialog;
import com.github.retrooper.packetevents.protocol.dialog.DialogListDialog;
import com.github.retrooper.packetevents.protocol.dialog.action.DynamicCustomAction;
import com.github.retrooper.packetevents.protocol.dialog.action.DynamicRunCommandAction;
import com.github.retrooper.packetevents.protocol.dialog.action.DialogTemplate;
import com.github.retrooper.packetevents.protocol.dialog.action.StaticAction;
import com.github.retrooper.packetevents.protocol.dialog.body.ItemDialogBody;
import com.github.retrooper.packetevents.protocol.dialog.body.PlainMessage;
import com.github.retrooper.packetevents.protocol.dialog.body.PlainMessageDialogBody;
import com.github.retrooper.packetevents.protocol.dialog.button.ActionButton;
import com.github.retrooper.packetevents.protocol.dialog.button.CommonButtonData;
import com.github.retrooper.packetevents.protocol.dialog.input.*;
import com.github.retrooper.packetevents.protocol.item.ItemStack;
import com.github.retrooper.packetevents.protocol.mapper.MappedEntitySet;
import com.github.retrooper.packetevents.protocol.nbt.*;
import com.github.retrooper.packetevents.protocol.player.ClientVersion;
import com.github.retrooper.packetevents.resources.ResourceLocation;
import ru.privatenull.pnauth.dialog.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** Lossless mapping from pnAuth dialogs to PacketEvents' vanilla protocol model. */
public final class PacketEventsDialogMapper {
    private PacketEventsDialogMapper() { }

    public static Dialog map(PlayerDialog source) {
        CommonDialogData common = common(source);
        if (source.type() instanceof DialogType.Notice notice) {
            return new NoticeDialog(common, button(notice.action()));
        }
        if (source.type() instanceof DialogType.Confirmation confirmation) {
            return new ConfirmationDialog(common, button(confirmation.yes()), button(confirmation.no()));
        }
        if (source.type() instanceof DialogType.MultiAction multi) {
            return new MultiActionDialog(common, buttons(multi.actions()), nullableButton(multi.exitAction()), multi.columns());
        }
        if (source.type() instanceof DialogType.ServerLinks links) {
            return new ServerLinksDialog(common, nullableButton(links.exitAction()), links.columns(), links.buttonWidth());
        }
        DialogType.DialogList list = (DialogType.DialogList) source.type();
        var references = list.dialogTag() == null
                ? new MappedEntitySet<>(list.dialogs().stream().map(PacketEventsDialogMapper::map).toList())
                : new MappedEntitySet<Dialog>(new ResourceLocation(list.dialogTag()));
        return new DialogListDialog(common, references, nullableButton(list.exitAction()),
                list.columns(), list.buttonWidth());
    }

    private static CommonDialogData common(PlayerDialog source) {
        var layout = source.layout();
        return new CommonDialogData(layout.title(), layout.externalTitle(), layout.canCloseWithEscape(),
                layout.pause(), com.github.retrooper.packetevents.protocol.dialog.DialogAction.valueOf(
                        layout.afterAction().name()), bodies(layout.body()), inputs(layout.inputs()));
    }

    private static List<com.github.retrooper.packetevents.protocol.dialog.body.DialogBody> bodies(
            List<DialogBody> source) {
        List<com.github.retrooper.packetevents.protocol.dialog.body.DialogBody> result = new ArrayList<>();
        for (DialogBody body : source) {
            if (body instanceof DialogBody.PlainMessage message) {
                result.add(new PlainMessageDialogBody(new PlainMessage(message.content(), message.width())));
            } else if (body instanceof DialogBody.Item item) {
                PlainMessage description = item.description() == null ? null
                        : new PlainMessage(item.description(), item.descriptionWidth());
                result.add(new ItemDialogBody(ItemStack.decode(compound(item.itemStack()), ClientVersion.V_1_21_6),
                        description, item.showDecorations(), item.showTooltip(), item.width(), item.height()));
            }
        }
        return List.copyOf(result);
    }

    private static List<Input> inputs(List<DialogInput> source) {
        List<Input> result = new ArrayList<>();
        for (DialogInput input : source) {
            InputControl control;
            if (input instanceof DialogInput.Text text) {
                var multiline = text.multiline() == null ? null : new TextInputControl.MultilineOptions(
                        text.multiline().maximumLines(), text.multiline().height());
                control = new TextInputControl(text.width(), text.label(), text.labelVisible(),
                        text.initialValue(), text.maximumLength(), multiline);
            } else if (input instanceof DialogInput.Toggle toggle) {
                control = new BooleanInputControl(toggle.label(), toggle.initialValue(),
                        toggle.onTrue(), toggle.onFalse());
            } else if (input instanceof DialogInput.Choice choice) {
                List<SingleOptionInputControl.Entry> options = choice.options().stream()
                        .map(option -> new SingleOptionInputControl.Entry(
                                option.id(), option.display(), option.initial())).toList();
                control = new SingleOptionInputControl(choice.width(), options,
                        choice.label(), choice.labelVisible());
            } else {
                DialogInput.NumberRange range = (DialogInput.NumberRange) input;
                control = new NumberRangeInputControl(range.width(), range.label(), range.labelFormat(),
                        new NumberRangeInputControl.RangeInfo(
                                range.start(), range.end(), range.initial(), range.step()));
            }
            result.add(new Input(input.id(), control));
        }
        return List.copyOf(result);
    }

    private static List<ActionButton> buttons(List<DialogButton> source) {
        return source.stream().map(PacketEventsDialogMapper::button).toList();
    }

    private static ActionButton nullableButton(DialogButton source) {
        return source == null ? null : button(source);
    }

    private static ActionButton button(DialogButton source) {
        return new ActionButton(new CommonButtonData(source.label(), source.tooltip(), source.width()),
                action(source.action()));
    }

    private static com.github.retrooper.packetevents.protocol.dialog.action.Action action(DialogAction source) {
        if (source == null || source instanceof DialogAction.None) return null;
        if (source instanceof DialogAction.DynamicRunCommand command) {
            return new DynamicRunCommandAction(new DialogTemplate(command.template()));
        }
        if (source instanceof DialogAction.DynamicCustom custom) {
            return new DynamicCustomAction(new ResourceLocation(custom.id()), compound(custom.additions()));
        }
        DialogAction.Static fixed = (DialogAction.Static) source;
        return new StaticAction(click(fixed.type(), fixed.payload()));
    }

    private static ClickEvent click(String rawType, Map<String, Object> payload) {
        String type = rawType.replace("minecraft:", "");
        return switch (type) {
            case "open_url" -> new OpenUrlClickEvent(string(payload, "url"));
            case "run_command" -> new RunCommandClickEvent(string(payload, "command"));
            case "suggest_command" -> new SuggestCommandClickEvent(string(payload, "command"));
            case "change_page" -> new ChangePageClickEvent(number(payload, "page").intValue());
            case "copy_to_clipboard" -> new CopyToClipboardClickEvent(string(payload, "value"));
            case "custom" -> new CustomClickEvent(new ResourceLocation(string(payload, "id")), nbt(payload.get("payload")));
            case "show_dialog" -> new ShowDialogClickEvent(map((PlayerDialog) payload.get("dialog")));
            default -> throw new IllegalArgumentException("Unsupported vanilla dialog action: " + rawType);
        };
    }

    private static NBTCompound compound(Map<String, Object> values) {
        NBTCompound result = new NBTCompound();
        values.forEach((key, value) -> result.setTag(key, nbt(value)));
        return result;
    }

    private static NBT nbt(Object value) {
        if (value == null) return new NBTCompound();
        if (value instanceof NBT existing) return existing;
        if (value instanceof Map<?, ?> map) {
            NBTCompound result = new NBTCompound();
            map.forEach((key, child) -> result.setTag(String.valueOf(key), nbt(child)));
            return result;
        }
        if (value instanceof List<?> list) {
            List<NBT> values = list.stream().map(PacketEventsDialogMapper::nbt).toList();
            @SuppressWarnings({"rawtypes", "unchecked"})
            NBTList<NBT> result = new NBTList((com.github.retrooper.packetevents.protocol.nbt.NBTType)
                    NBTList.getCommonTagType(values), values);
            return result;
        }
        if (value instanceof Boolean bool) return new NBTByte((byte) (bool ? 1 : 0));
        if (value instanceof Byte number) return new NBTByte(number);
        if (value instanceof Short number) return new NBTShort(number);
        if (value instanceof Integer number) return new NBTInt(number);
        if (value instanceof Long number) return new NBTLong(number);
        if (value instanceof Float number) return new NBTFloat(number);
        if (value instanceof Double number) return new NBTDouble(number);
        return new NBTString(String.valueOf(value));
    }

    private static String string(Map<String, Object> payload, String key) {
        Object value = payload.get(key);
        if (value == null) throw new IllegalArgumentException("Missing action field: " + key);
        return String.valueOf(value);
    }

    private static Number number(Map<String, Object> payload, String key) {
        Object value = payload.get(key);
        if (value instanceof Number number) return number;
        throw new IllegalArgumentException("Action field is not numeric: " + key);
    }
}

File: plugin/src/main/kotlin/ru/privatenull/pnauth/transport/packetevents/PacketEventsDialogMapper.kt
package ru.privatenull.pnauth.transport.packetevents

import com.github.retrooper.packetevents.protocol.chat.clickevent.ChangePageClickEvent
import com.github.retrooper.packetevents.protocol.chat.clickevent.ClickEvent
import com.github.retrooper.packetevents.protocol.chat.clickevent.CopyToClipboardClickEvent
import com.github.retrooper.packetevents.protocol.chat.clickevent.CustomClickEvent
import com.github.retrooper.packetevents.protocol.chat.clickevent.OpenUrlClickEvent
import com.github.retrooper.packetevents.protocol.chat.clickevent.RunCommandClickEvent
import com.github.retrooper.packetevents.protocol.chat.clickevent.ShowDialogClickEvent
import com.github.retrooper.packetevents.protocol.chat.clickevent.SuggestCommandClickEvent
import com.github.retrooper.packetevents.protocol.dialog.CommonDialogData
import com.github.retrooper.packetevents.protocol.dialog.ConfirmationDialog
import com.github.retrooper.packetevents.protocol.dialog.Dialog
import com.github.retrooper.packetevents.protocol.dialog.DialogListDialog
import com.github.retrooper.packetevents.protocol.dialog.MultiActionDialog
import com.github.retrooper.packetevents.protocol.dialog.NoticeDialog
import com.github.retrooper.packetevents.protocol.dialog.ServerLinksDialog
import com.github.retrooper.packetevents.protocol.dialog.action.Action
import com.github.retrooper.packetevents.protocol.dialog.action.DialogTemplate
import com.github.retrooper.packetevents.protocol.dialog.action.DynamicCustomAction
import com.github.retrooper.packetevents.protocol.dialog.action.DynamicRunCommandAction
import com.github.retrooper.packetevents.protocol.dialog.action.StaticAction
import com.github.retrooper.packetevents.protocol.dialog.body.ItemDialogBody
import com.github.retrooper.packetevents.protocol.dialog.body.PlainMessage
import com.github.retrooper.packetevents.protocol.dialog.body.PlainMessageDialogBody
import com.github.retrooper.packetevents.protocol.dialog.button.ActionButton
import com.github.retrooper.packetevents.protocol.dialog.button.CommonButtonData
import com.github.retrooper.packetevents.protocol.dialog.input.BooleanInputControl
import com.github.retrooper.packetevents.protocol.dialog.input.Input
import com.github.retrooper.packetevents.protocol.dialog.input.InputControl
import com.github.retrooper.packetevents.protocol.dialog.input.NumberRangeInputControl
import com.github.retrooper.packetevents.protocol.dialog.input.SingleOptionInputControl
import com.github.retrooper.packetevents.protocol.dialog.input.TextInputControl
import com.github.retrooper.packetevents.protocol.item.ItemStack
import com.github.retrooper.packetevents.protocol.mapper.MappedEntitySet
import com.github.retrooper.packetevents.protocol.nbt.NBT
import com.github.retrooper.packetevents.protocol.nbt.NBTByte
import com.github.retrooper.packetevents.protocol.nbt.NBTCompound
import com.github.retrooper.packetevents.protocol.nbt.NBTDouble
import com.github.retrooper.packetevents.protocol.nbt.NBTFloat
import com.github.retrooper.packetevents.protocol.nbt.NBTInt
import com.github.retrooper.packetevents.protocol.nbt.NBTList
import com.github.retrooper.packetevents.protocol.nbt.NBTLong
import com.github.retrooper.packetevents.protocol.nbt.NBTNumber
import com.github.retrooper.packetevents.protocol.nbt.NBTShort
import com.github.retrooper.packetevents.protocol.nbt.NBTString
import com.github.retrooper.packetevents.protocol.nbt.NBTType
import com.github.retrooper.packetevents.protocol.player.ClientVersion
import com.github.retrooper.packetevents.resources.ResourceLocation
import ru.privatenull.pnauth.dialog.DialogAction
import ru.privatenull.pnauth.dialog.DialogBody
import ru.privatenull.pnauth.dialog.DialogButton
import ru.privatenull.pnauth.dialog.DialogInput
import ru.privatenull.pnauth.dialog.DialogType
import ru.privatenull.pnauth.dialog.PlayerDialog

/** Lossless mapping from pnAuth dialogs to PacketEvents' vanilla protocol model. */
object PacketEventsDialogMapper {

    @JvmStatic
    fun map(source: PlayerDialog): Dialog {
        val common = common(source)
        return when (val type = source.type) {
            is DialogType.Notice -> NoticeDialog(common, button(type.action))
            is DialogType.Confirmation -> ConfirmationDialog(common, button(type.yes), button(type.no))
            is DialogType.MultiAction -> MultiActionDialog(
                common,
                buttons(type.actions),
                nullableButton(type.exitAction),
                type.columns
            )
            is DialogType.ServerLinks -> ServerLinksDialog(
                common,
                nullableButton(type.exitAction),
                type.columns,
                type.buttonWidth
            )
            is DialogType.DialogList -> {
                val references = if (type.dialogTag == null) {
                    MappedEntitySet(type.dialogs.map { map(it) })
                } else {
                    MappedEntitySet<Dialog>(ResourceLocation(type.dialogTag!!))
                }
                DialogListDialog(
                    common,
                    references,
                    nullableButton(type.exitAction),
                    type.columns,
                    type.buttonWidth
                )
            }
        }
    }

    private fun common(source: PlayerDialog): CommonDialogData {
        val layout = source.layout
        return CommonDialogData(
            layout.title,
            layout.externalTitle,
            layout.canCloseWithEscape,
            layout.pause,
            com.github.retrooper.packetevents.protocol.dialog.DialogAction.valueOf(layout.afterAction.name),
            bodies(layout.body),
            inputs(layout.inputs)
        )
    }

    private fun bodies(source: List<DialogBody>): List<com.github.retrooper.packetevents.protocol.dialog.body.DialogBody> {
        val result = mutableListOf<com.github.retrooper.packetevents.protocol.dialog.body.DialogBody>()
        for (body in source) {
            when (body) {
                is DialogBody.PlainMessage -> {
                    result.add(PlainMessageDialogBody(PlainMessage(body.content, body.width)))
                }
                is DialogBody.Item -> {
                    val description = body.description?.let { PlainMessage(it, body.descriptionWidth) }
                    @Suppress("DEPRECATION")
                    val item = ItemStack.decode(compound(body.itemStack), ClientVersion.V_1_21_6)
                    result.add(
                        ItemDialogBody(
                            item,
                            description,
                            body.showDecorations,
                            body.showTooltip,
                            body.width,
                            body.height
                        )
                    )
                }
            }
        }
        return result.toList()
    }

    private fun inputs(source: List<DialogInput>): List<Input> {
        val result = mutableListOf<Input>()
        for (input in source) {
            val control: InputControl = when (input) {
                is DialogInput.Text -> {
                    val multiline = if (input.multiline == null) null
                    else TextInputControl.MultilineOptions(
                        input.multiline!!.maximumLines,
                        input.multiline!!.height
                    )
                    TextInputControl(
                        input.width,
                        input.label(),
                        input.labelVisible,
                        input.initialValue,
                        input.maximumLength,
                        multiline
                    )
                }
                is DialogInput.Toggle -> {
                    BooleanInputControl(
                        input.label(),
                        input.initialValue,
                        input.onTrue,
                        input.onFalse
                    )
                }
                is DialogInput.Choice -> {
                    val options = input.options.map { option ->
                        SingleOptionInputControl.Entry(option.id, option.display, option.initial)
                    }
                    SingleOptionInputControl(
                        input.width,
                        options,
                        input.label(),
                        input.labelVisible
                    )
                }
                is DialogInput.NumberRange -> {
                    NumberRangeInputControl(
                        input.width,
                        input.label(),
                        input.labelFormat,
                        NumberRangeInputControl.RangeInfo(
                            input.start,
                            input.end,
                            input.initial,
                            input.step
                        )
                    )
                }
            }
            result.add(Input(input.id(), control))
        }
        return result.toList()
    }

    private fun buttons(source: List<DialogButton>): List<ActionButton> {
        return source.map { button(it) }
    }

    private fun nullableButton(source: DialogButton?): ActionButton? {
        return source?.let { button(it) }
    }

    private fun button(source: DialogButton): ActionButton {
        return ActionButton(
            CommonButtonData(source.label, source.tooltip, source.width),
            action(source.action)
        )
    }

    private fun action(source: DialogAction?): Action? {
        if (source == null || source is DialogAction.None) return null
        if (source is DialogAction.DynamicRunCommand) {
            return DynamicRunCommandAction(DialogTemplate(source.template))
        }
        if (source is DialogAction.DynamicCustom) {
            return DynamicCustomAction(ResourceLocation(source.id), compound(source.additions))
        }
        val fixed = source as DialogAction.Static
        return StaticAction(click(fixed.type, fixed.payload))
    }

    private fun click(rawType: String, payload: Map<String, Any>): ClickEvent {
        val type = rawType.replace("minecraft:", "")
        return when (type) {
            "open_url" -> OpenUrlClickEvent(string(payload, "url"))
            "run_command" -> RunCommandClickEvent(string(payload, "command"))
            "suggest_command" -> SuggestCommandClickEvent(string(payload, "command"))
            "change_page" -> ChangePageClickEvent(number(payload, "page").toInt())
            "copy_to_clipboard" -> CopyToClipboardClickEvent(string(payload, "value"))
            "custom" -> CustomClickEvent(ResourceLocation(string(payload, "id")), nbt(payload["payload"]))
            "show_dialog" -> ShowDialogClickEvent(map(payload["dialog"] as PlayerDialog))
            else -> throw IllegalArgumentException("Unsupported vanilla dialog action: $rawType")
        }
    }

    private fun compound(values: Map<String, Any?>): NBTCompound {
        val result = NBTCompound()
        values.forEach { (key, value) -> result.setTag(key, nbt(value)) }
        return result
    }

    private fun nbt(value: Any?): NBT {
        if (value == null) return NBTCompound()
        if (value is NBT) return value
        if (value is Map<*, *>) {
            val result = NBTCompound()
            value.forEach { (key, child) -> result.setTag(key.toString(), nbt(child)) }
            return result
        }
        if (value is List<*>) {
            val values = value.map { nbt(it) }
            @Suppress("UNCHECKED_CAST")
            return NBTList(NBTList.getCommonTagType(values) as NBTType<NBT>, values)
        }
        if (value is Boolean) return NBTByte(if (value) 1.toByte() else 0.toByte())
        if (value is Byte) return NBTByte(value)
        if (value is Short) return NBTShort(value)
        if (value is Int) return NBTInt(value)
        if (value is Long) return NBTLong(value)
        if (value is Float) return NBTFloat(value)
        if (value is Double) return NBTDouble(value)
        return NBTString(value.toString())
    }

    private fun string(payload: Map<String, Any>, key: String): String {
        val value = payload[key] ?: throw IllegalArgumentException("Missing action field: $key")
        return value.toString()
    }

    private fun number(payload: Map<String, Any>, key: String): Number {
        val value = payload[key]
        if (value is Number) return value
        throw IllegalArgumentException("Action field is not numeric: $key")
    }
}
